packages/connect/src/core/onCallFirmwareUpdate.ts

```diff
@@ -0,0 +1,484 @@
+    );
+
+    if (!reconnectedDevice) {
+        throw ERRORS.TypedError('Method_Interrupted');
+    }
+
+    registerEvents(reconnectedDevice, postMessage);
+    await reconnectedDevice.waitForFirstRun();
+    await reconnectedDevice.acquire();
+
+    return reconnectedDevice;
+};
+
+const waitForDisconnectedDevice = async ({
+    params: { manual },
+    context: { deviceList, device, postMessage, log },
+}: {
+    params: { manual: boolean };
+    context: {
+        deviceList: DeviceList;
+        device: Device;
+        postMessage: PostMessage;
+        log: Log;
+    };
+}): Promise<void> => {
+    log.debug('onCallFirmwareUpdate', 'waiting for device to disconnect');
+    postMessage(
+        createUiMessage(UI.FIRMWARE_DISCONNECT, {
+            device: device.toMessageObject(),
+            manual,
+        }),
+    );
+    await new Promise(resolve => {
+        deviceList.once('device-disconnect', resolve);
+    });
+};
+
+const getInstallationParams = (device: Device, binary?: ArrayBuffer) => {
+    // we can detect support properly only if device was not connected in bootloader mode
+    if (!device.features.bootloader_mode) {
+        const version = binary ? parseFirmwareHeaders(Buffer.from(binary)).version : undefined;
+        const isUpdatingToNewerVersion = !version
+            ? device.firmwareRelease?.isNewer
+            : isNewer(version, [
+                  device.features.major_version,
+                  device.features.minor_version,
+                  device.features.patch_version,
+              ]);
+
+        const upgrade = device.atLeast('2.6.3') && isUpdatingToNewerVersion;
+        const manual = !device.atLeast(['1.10.0', '2.6.0']) && !upgrade;
+        const language = device.atLeast('2.7.0');
+
+        return {
+            /** RebootToBootloader is not supported */
+            manual,
+            /** RebootToBootloader (REBOOT_AND_UPGRADE) is supported  */
+            upgrade,
+            /** Language update is supported */
+            language,
+        };
+    } else {
+        // if device connected initially in bootloader mode:
+        // manual: false - device is already in bootloader, so this field doesn't matter
+        // upgrade,language: false - we don't know if supported, so take the safest route and don't use these features
+        return {
+            manual: false,
+            upgrade: false,
+            language: false,
+        };
+    }
+};
+
+const getFwHeader = (binary: ArrayBuffer) => Buffer.from(binary.slice(0, 6000)).toString('hex');
+
+const getBinaryHelper = (
+    device: Device,
+    params: Params,
+    log: Log,
+    postMessage: PostMessage,
+    intermediaryVersion?: IntermediaryVersion,
+) => {
+    if (!device.firmwareRelease) {
+        throw ERRORS.TypedError('Runtime', 'device.firmwareRelease is not set');
+    }
+    const btcOnly = params.btcOnly || device.firmwareType === 'bitcoin-only';
```

Comment:
```suggestion
    const btcOnly = params.btcOnly || (params.btcOnly === undefined && device.firmwareType === 'bitcoin-only');
```

Reply:
This way, you can't switch from bitcoin-only to regular.

Reply:
thanks. makes sense

Reply:
8d7bd1d680355a9b7d6219df7fdcb1273e286acd